langchain/src/agents/openai/output_parser.ts

```diff
@@ -0,0 +1,83 @@
+import type { OpenAI as OpenAIClient } from "openai";
+import {
+  AgentAction,
+  AgentFinish,
+  BaseMessage,
+  ChatGeneration,
+  Generation,
+  isBaseMessage,
+} from "../../schema/index.js";
+import { AgentActionOutputParser } from "../types.js";
+import { OutputParserException } from "../../schema/output_parser.js";
+
+/**
+ * Type that represents an agent action with an optional message log.
+ */
+export type FunctionsAgentAction = AgentAction & {
+  messageLog?: BaseMessage[];
+};
+
+export class OpenAIFunctionsAgentOutputParser extends AgentActionOutputParser {
+  lc_namespace = ["langchain", "agents", "openai"];
+
+  constructor() {
+    super(...arguments);
+  }
+
+  async parse(text: string): Promise<AgentAction | AgentFinish> {
+    throw new Error(
+      `OpenAIFunctionsAgentOutputParser can only parse messages.\nPassed input: ${text}`
+    );
+  }
+
+  async parseResult(generations: Generation[] | ChatGeneration[]) {
+    if ("message" in generations[0] && isBaseMessage(generations[0].message)) {
+      return this.parseAIMessage(generations[0].message);
+    }
+    throw new Error(
+      "parseResult on OpenAIFunctionsAgentOutputParser only works on ChatGeneration output"
+    );
+  }
+
+  /**
+   * Parses the output message into a FunctionsAgentAction or AgentFinish
+   * object.
+   * @param message The BaseMessage to parse.
+   * @returns A FunctionsAgentAction or AgentFinish object.
+   */
+  parseAIMessage(message: BaseMessage): FunctionsAgentAction | AgentFinish {
```

Comment:
Fine to have separate but could we just put this in `parseResult`?

Reply:
Yeah, refactoring to do this, had two because that's what they have in py

Reply:
Ah never mind, need this because the message response from the `_call` method in openai functions isn't of type `ChatGeneration`